lib/vtls/gskit.c

```diff
@@ -978,9 +979,8 @@ static CURLcode gskit_connect_step2(struct Curl_easy *data,
     timediff_t timeout_ms = nonblocking? 0: Curl_timeleft(data, NULL, TRUE);
     if(timeout_ms < 0)
       timeout_ms = 0;
-    stmv.tv_sec = timeout_ms / 1000;
-    stmv.tv_usec = (timeout_ms - stmv.tv_sec * 1000) * 1000;
-    switch(QsoWaitForIOCompletion(BACKEND->iocport, &cstat, &stmv)) {
+    switch(QsoWaitForIOCompletion(BACKEND->iocport, &cstat,
+                                  curlx_mstotv(timeout_ms))) {
```

Comment:
this one doesn't look right

Reply:
maybe

~~~c
    curlx_mstotv(&stmv, timeout_ms);
    switch(QsoWaitForIOCompletion(BACKEND->iocport, &cstat, &smtv)) {
~~~

?

Reply:
hm, to me it looks like using `select`-semantics for this parameter is just fine according to IBM documentation:

> **struct timeval * timeToWait**
> (Input) A pointer to a timeval structure that contains the time in seconds and microseconds for which the QsoWaitForIOCompletion() call should block if there is no completion status to receive.
> _If this parameter is null, QsoWaitForIOCompletion() waits indefinitely. If this value is specified, and 0 seconds 0 microseconds are specified, QsoWaitForIOCompletion() returns immediately._

curlx_mstotv aligns exactly do that 2nd sentence by returning NULL if stmv is also NULL (or timeout_ms is < 0) and setting the timeval members to 0 if timeout_ms is 0.

Reply:
curlx_mstotv takes two parameters not one. also I don't know why we'd want to wait indefinitely that doesn't make sense

Reply:
You are right, sorry, I totally missed the 2nd parameter thing. If timeout_ms is actually less then 0 and curlx_mstotv returns NULL, then smtv would be uninitialized and therefore the behaviour would still be undefined anyway.

Reply:
Actually timeout_ms cannot be less then 0 because of this preceding code:
```
    if(timeout_ms < 0)
      timeout_ms = 0;
```
So we won't ever block as curlx_mstotv will never return NULL in this case.